File: algorithm/sorting.insertion.mjs
import assert from "node:assert";

function sort(arr) {
  for (let i = 1; i < arr.length; i++) {
    let targetIndex = i;
    let targetValue = arr[i];
    for (let j = i - 1; j >= 0; j -= 1) {
      if (arr[j] > targetValue) {
        targetIndex = j;
        arr[j + 1] = arr[j];
      } else {
        break;
      }
    }
    arr[targetIndex] = targetValue;
  }
  return arr;
}

let given = [5, 7, 8, 1, 4, 6, 9];
let expected = [1, 4, 5, 6, 7, 8, 9];
let output = sort(given);
assert.deepEqual(output, expected);

File: algorithm/sorting.insertion.py
# for i in range(5, -1, -1):
#     print('iiii: ', i)


def sort(arr):
    for i in range(1, len(arr)):
        pIndex, pValue = i, arr[i]
        for j in range(i-1, -1, -1):
            if pValue < arr[j]:
                arr[j+1] = arr[j]
                pIndex = j
            else:
                break
        arr[pIndex] = pValue
    return arr


given = [5, 7, 8, 1, 4, 6, 9]
expected = [1, 4, 5, 6, 7, 8, 9]
output = sort(given)
print(output)
